File: modules/client/rooms/send.cc
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.

#include "rooms.h"

using namespace ircd::m;
using namespace ircd;

static bool
check_transaction_id(const m::user::id &user_id,
                     const string_view &transaction_id);

static void
save_transaction_id(const m::event &,
                    m::vm::eval &);

static m::event::id::buf
handle_command(client &,
               const m::resource::request &,
               const m::room &,
               const string_view & = {});

m::hookfn<m::vm::eval &>
save_transaction_id_hookfn
{
	save_transaction_id,
	{
		{ "_site",    "vm.post" },
		{ "origin",   my_host() },
	}
};

conf::item<bool>
new_content_workaround
{
	{ "name",     "ircd.client.rooms.send.new_content_workaround" },
	{ "default",  true                                            },
};

static_assert
(
	m::event::MAX_SIZE >= 1_KiB
);

static const size_t
content_max
{
	m::event::MAX_SIZE - 1_KiB
};

m::resource::response
put__send(client &client,
          const m::resource::request &request,
          const room::id &room_id)
{
	if(request.parv.size() < 3)
		throw NEED_MORE_PARAMS
		{
			"type parameter missing"
		};

	char type_buf[m::event::TYPE_MAX_SIZE];
	const string_view &type
	{
		url::decode(type_buf, request.parv[2])
	};

	if(request.parv.size() < 4)
		throw NEED_MORE_PARAMS
		{
			"txnid parameter missing"
		};

	char transaction_id_buf[64];
	const string_view &transaction_id
	{
		url::decode(transaction_id_buf, request.parv[3])
	};

	if(!check_transaction_id(request.user_id, transaction_id))
		throw m::error
		{
			http::CONFLICT, "M_DUPLICATE_TXNID",
			"Already processed request with txnid '%s'",
			transaction_id
		};

	json::object content
	{
		request
	};

	// Workaround for the quadruplication of content by certain clients
	// supporting message edits via `m.new_content`. The skinny is that
	// there's effectively four copies of the user's message data which can
	// be found within the content here. For this reason if the content size
	// with all four copies will exceed the threshold, we truncate two of the
	// four copies which won't be used by supporting clients.
	const bool workaround_new_content
	{
		// This functionality is enabled by the configuration
		new_content_workaround

		// For good-faith compatibility this functionality is enabled iff the
		// content size is excessive.
		&& size(string_view(content)) > content_max

		// This workaround is only effective for m.new_content edits.
		&& content.has("m.new_content")
	};

	json::strung shorter_content;
	if(workaround_new_content)
		content = shorter_content = json::replace(content, json::members
		{
			{ "body",            " * "_sv  },
			{ "formatted_body",  " * "_sv  },
		});

	// This is only a preliminary check that the content size is sane and
	// will fit. There may still be a rejection at a deeper stage.
	if(size(string_view(content)) > content_max)
		throw m::error
		{
			http::PAYLOAD_TOO_LARGE, "M_TOO_LARGE",
			"Message of %zu bytes exceeds maximum of %zu bytes.",
			size(request.content),
			content_max,
		};

	m::vm::copts copts;
	copts.client_txnid = transaction_id;
	const room room
	{
		room_id, &copts
	};

	bool command_echo {false};
	if(type == "m.room.message")
	{
		const m::room::message message
		{
			content
		};

		const bool command
		{
			json::get<"msgtype"_>(message) == "m.text"
			&& startswith(json::get<"body"_>(message), "\\\\")
		};

		const bool echo
		{
			startswith(lstrip(json::get<"body"_>(message), "\\\\"), '!')
		};

		if(command && !echo)
		{
			const auto event_id
			{
				handle_command(client, request, room)
			};

			return m::resource::response
			{
				client, json::members
				{
					{ "event_id",  event_id },
					{ "cmd",       true     },
				}
			};
		}
		else command_echo = command && echo;
	}

	const auto event_id
	{
		m::send(room, request.user_id, type, content)
	};

	// For public echo, run the command after sending the command to the room.
	if(command_echo)
	{
		const auto cmd_event_id
		{
			handle_command(client, request, room, event_id)
		};
	}

	return m::resource::response
	{
		client, json::members
		{
			{ "event_id", event_id }
		}
	};
}

m::event::id::buf
handle_command(client &client,
               const m::resource::request &request,
               const m::room &room,
               const string_view &echo_id)
{
	const user::room user_room
	{
		request.user_id, room.copts
	};

	const auto event_id
	{
		send(user_room, request.user_id, "ircd.cmd",
		{
			{ "msgtype",  "m.text"         },
			{ "body",     request["body"]  },
			{ "room_id",  room.room_id     },
			{ "event_id", echo_id          },
		})
	};

	return event_id;
}

void
save_transaction_id(const m::event &event,
                    m::vm::eval &eval)
{
	if(!eval.copts)
		return;

	if(!eval.copts->client_txnid)
		return;

	if(!event.event_id)
		return;

	assert(my_host(at<"origin"_>(event)));
	const m::user::room user_room
	{
		at<"sender"_>(event)
	};

	static const string_view &type
	{
		"ircd.client.txnid"
	};

	const auto &state_key
	{
		event.event_id
	};

	send(user_room, at<"sender"_>(event), type, state_key,
	{
		{ "transaction_id", eval.copts->client_txnid }
	});
}

// Using a linear search here because we have no index on txnids as this
// is the only codepath where we'd perform that lookup; in contrast the
// event_id -> txnid query is made far more often for client sync.
//
// This means we have to set some arbitrary limits on the linear search:
// lim[0] is a total limit of events to iterate, so if the user's room
// has a lot of activity we might return a false non-match and allow a
// duplicate txnid; this is highly unlikely. lim[1] allows the user to
// have several /sends in flight at the same time, also unlikely but we
// avoid that case for false non-match.
static bool
check_transaction_id(const m::user::id &user_id,
                     const string_view &transaction_id)
{
	static const auto type_match
	{
		[](const string_view &type)
		{
			return type == "ircd.client.txnid";
		}
	};

	const auto content_match
	{
		[&transaction_id](const json::object &content)
		{
			const json::string &value
			{
				content["transaction_id"]
			};

			return value == transaction_id;
		}
	};

	ssize_t lim[] { 128, 3 };
	const m::user::room user_room{user_id};
	for(m::room::events it(user_room); it && lim[0] > 0 && lim[1] > 0; --it, --lim[0])
	{
		if(!m::query(std::nothrow, it.event_idx(), "type", type_match))
			continue;

		--lim[1];
		if(!m::query(std::nothrow, it.event_idx(), "content", content_match))
			continue;

		return false;
	}

	return true;
}

File: share/webapp/js/room/send.js
/*
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.
*/

'use strict';


/**************************************
 * Input send interface.
 *
 * Handles the transmission process for input.
 */

/**
 * Any text from the input area which is intended for transmission lands here
 * and is dispatched. This function may be skipped to call another send
 * sub-method directly.
 */
room.prototype.send = function(text)
{
	let emote_prefix = "/me ";
	if(text.startsWith(emote_prefix))
	{
		let msg = text.slice(emote_prefix.length);
		return room.send.message.emote.call(this, msg);
	}

	if(text.endsWith(".jpg") || text.endsWith(".png"))
		return room.send.message.image.call(this, text);

	if(text.startsWith("https://www.youtube.com/watch?v="))
	{
		let a = text.split(" ", 2);
		let url = a[0];
		let desc = a[1];
		return room.send.message.video.call(this, url, desc);
	}

	return room.send.message.text.call(this, text);
};

/**
 * TODO
 */
room.prototype.restate = function($event, type)
{
	return room.send.state.call(this, type);
};

/**
 * The composed event content is transmitted out of here.
 * Should not be called directly. Use a helper below.
 *
 * this = room
 */
room.send = function(type, opts = {})
{
	let handler = (error, data) => callback(arguments, error, data);
	return client.m.rooms.send.put(this.id, type, opts, handler);
};

/**
 * TODO
 */
room.send.state = function(type, state_key = "")
{
	let event = this.state[type];
	if(state_key.length > 0)
		event = event[state_key];

	let pending = this.pending[type];
	if(state_key.length > 0)
		pending = pending[state_key];

	if(pending.content === undefined)
		return;

	let opts = {};
	return client.m.rooms.state.put(this.id, type, state_key, pending.content, opts, (error, data) =>
	{
		if(error)
			throw new client.error(error);

		pending.event_id = data.event_id;
		pending.type = type;
		pending.state_key = state_key;
		pending.sender = client.session.user_id;

		this.timeline.insert(pending);
		if(state_key.length > 0)
			delete this.pending[type][state_key];
		else
			delete this.pending[type];
	});
};

/**
 * The composed m.room.message event content is transmitted out of here.
 * Should not be called directly. Use or create a helper below.
 *
 * this = room
 */
room.send.message = function(opts = {})
{
	let handler = (error, data) =>
	{
		if(error !== undefined)
			this.scroll.to.bottom("fast");

		return callback(arguments, error, data);
	};

	return room.send.call(this, "m.room.message", opts, handler);
};

/** Compose a text message
 */
room.send.message.text = function(text, opts = {})
{
	Object.update(opts,
	{
		content:
		{
			body: text,
			msgtype: "m.text",
		},
	});

	let handler = (error, data) => callback(arguments, error, data);
	return room.send.message.call(this, opts, handler);
};

/** Compose an emote message
 */
room.send.message.emote = function(text, opts = {})
{
	Object.update(opts,
	{
		content:
		{
			body: text,
			msgtype: "m.emote",
		}
	});

	let handler = (error, data) => callback(arguments, error, data);
	return room.send.message.call(this, opts, handler);
};

/** Compose an image message
 */
room.send.message.image = function(url, body = url, opts = {})
{
	Object.update(opts,
	{
		content:
		{
			body: body,
			msgtype: "m.image",
			url: url,
		},
	});

	let handler = (error, data) => callback(arguments, error, data);
	return room.send.message.call(this, opts, handler);
};

/** Compose a video message.
 */
room.send.message.video = function(url, body = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz", opts = {})
{
	Object.update(opts,
	{
		content:
		{
			body: body,
			info:
			{
				mimetype: "text/plain"
			},
			msgtype: "m.video",
			url: url,
		},
	});

	let handler = (error, data) => callback(arguments, error, data);
	return room.send.message.call(this, opts, handler);
};
